configserver/src/main/java/com/yahoo/vespa/config/server/session/SessionPreparer.java

```diff
@@ -225,6 +234,19 @@ void writeRotZK() {
             checkTimeout("write rotations to zookeeper");
         }
 
+        void writeContainerEndpointsZK(Optional<String> globalServiceId) {
+            if (!params.containerEndpoints().isEmpty()) { // Use endpoints from parameter when explicitly given
+                containerEndpoints.write(applicationId, params.containerEndpoints());
+            } else { // Fall back to writing rotations as container endpoints
+                if (globalServiceId.isEmpty()) {
+                    log.log(LogLevel.WARNING, "Want to write rotations " + rotationsSet + " as container endpoints, but " + applicationId + " has no global-service-id. This should not happen");
```

Comment:
I am not sure I understand the warning, not all applications have a global service id?

Reply:
No, it's an optional attribute in _deployment.xml_. If a tenant sets `global-service-id` and then removes it at a later time, this may happen as we never unassign rotations. The log message is here to verify if such a condition is present in main today.

Reply:
Sorry for the bad wording in the question. What I meant to ask was: If an application has no global service id and no rotations, why log here? I.e. shouldn't you only log when rotation set is empty?

Reply:
True, thanks!